### ISSUE TITLE
seq2seq model is not efficient when using sampled softmax

### ISSUE THREAD

**Issue by cloudygoose** [tensorflow/tensorflow#4138]
Hi, 
my commit number is
70de76e696c21da617fd2e6435cf7fedab220db8

I want to try sampled softmax to speed up my lm training. I'm changing slightly on the ptb training example code. I'm working on a CPU machine.
First, on ptb(vocab 10k), I'm using the "small" config, I see some speed up gain:
normal softmax 896wps

sample_softmax h256L1ba20 sample512 1975wps

Then, I move to a larger set, and using 100k vocab, I'm still setting the sample number to 512, I think the speed should be similar to the ptb case. But I got the speed to be about 300 wps, even if I set the sample number as small as 4. I don't understand why it should be so slower than the ptb case. Do you know how can I make it faster?

Here's some of the related code:

```
      output = tf.reshape(tf.concat(1, outputs), [-1, size])
      softmax_w = tf.get_variable("softmax_w", [size, vocab_size])
      softmax_w_t = tf.transpose(softmax_w)
      softmax_b = tf.get_variable("softmax_b", [vocab_size])

    if use_sample_softmax == True:
      loss = tf.nn.sampled_softmax_loss(softmax_w_t, softmax_b, output, tf.reshape(self._targets, [-1, 1]), num_samples, vocab_size) #todo
      self._cost = cost = tf.reduce_sum(loss) / batch_size
    else:
      logits = tf.matmul(output, softmax_w) + softmax_b
      loss = tf.nn.seq2seq.sequence_loss_by_example(
          [logits],
          [tf.reshape(self._targets, [-1])],
          [tf.ones([batch_size * num_steps])])
      self._cost = cost = tf.reduce_sum(loss) / batch_size


```

Thanks!
Goose

**Comment by mrry** [comment]
I suspect the issue is that the weight matrix is transposed, and this is causing the automatically-generated gradients code to take a slow path. In particular, `tf.nn.sampled_softmax_loss()` will generate sparse gradients for the weights, but the [gradient function for `tf.transpose()`](https://github.com/tensorflow/tensorflow/blob/9725395b8d68df30cfc81d0076db2365e49753fb/tensorflow/python/ops/array_grad.py#L341) doesn't have an optimized implementation for handling sparse `tf.IndexedSlices` objects.

Can you try running an experiment where you define the `softmax_w` variable as having shape `[vocab_size, size]` so that you don't have to transpose it? I would expect the performance of the sampled softmax to be much better in that case.

**Comment by cloudygoose** [comment]
Hi Mrry,

Thanks very much for your kind reply!
as you suggested, I changed the part of code to:
```
    output = tf.reshape(tf.concat(1, outputs), [-1, size])
    softmax_w_t = tf.get_variable("softmax_w", [vocab_size, size])
    softmax_w = tf.transpose(softmax_w_t)
    softmax_b = tf.get_variable("softmax_b", [vocab_size])

    if use_sample_softmax == True:
      loss = tf.nn.sampled_softmax_loss(softmax_w_t, softmax_b, output, tf.reshape(self._targets, [-1, 1]), num_samples, vocab_size) #todo
      self._cost = cost = tf.reduce_sum(loss) / batch_size
    else:
      logits = tf.matmul(output, softmax_w) + softmax_b
      loss = tf.nn.seq2seq.sequence_loss_by_example(
          [logits],
          [tf.reshape(self._targets, [-1])],
          [tf.ones([batch_size * num_steps])])
      self._cost = cost = tf.reduce_sum(loss) / batch_size
```
And yes, the speed now becomes
sample 
1024 1500wps 
128 1731wps

It seems more normal now.

Also, maybe the developers want to change the code in source/tensorflow/tensorflow/models/rnn/translate/seq2seq_model.py, because there a transpose of matrix is also feed into a smapled softmax

Thanks!
Goose

**Comment by mrry** [comment]
Thanks for the suggestion - assuming there are no knock-on effects to the full softmax case, we should fix the sample code to avoid the performance pitfall on sampled softmax. I'll assign this to the original author, @lukaszkaiser, in case there's a reason to avoid doing this, but this is probably an area where we'd appreciate contributions.

**Comment by lukaszkaiser** [comment]
Fix in #4270.

**Comment by vince62s** [comment]
@cloudygoose I am doing the same thing for a proper sampled_softmax performance.
it works fine for training.
However, a testing time, with num_steps=1 and bat_size=1, it make a huge loop with a tf.transpose for each iter (ie word) which makes it very slow.
How can we avoid to execute this transpose so many times at test time ?
